File: crates/se-mlclient/src/contract.rs
//! Request/response structs mirroring the Python `se_ml.contract` models EXACTLY.
//!
//! Field names are snake_case and identical to the pydantic models in
//! `ml-worker/src/se_ml/contract.py`. Bulk data (features + labels) is never embedded
//! here; it crosses the boundary as a Parquet file referenced by `dataset_uri`. These
//! bodies carry only metadata + metrics.
//!
//! `se-mlclient` is the ONLY crate in the workspace that knows the Python sidecar's
//! wire format. Everything downstream consumes the typed results, never raw JSON.

use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;

/// `GET /health` -> `{status, version}`.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct HealthResponse {
    pub status: String,
    pub version: String,
}

// --------------------------------------------------------------------------- //
// /fit
// --------------------------------------------------------------------------- //

/// `POST /fit` request body.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct MlFitRequest {
    /// Parquet path or `file://` URI for the labeled feature matrix.
    pub dataset_uri: String,
    /// Horizon profile id, e.g. `"swing"` (use [`se_core::Horizon::as_str`]).
    pub horizon: String,
    /// Free-form LightGBM hyperparameters; empty means the worker's defaults.
    #[serde(default)]
    pub model_params: serde_json::Map<String, serde_json::Value>,
    /// Deterministic seed.
    pub seed: i64,
}

/// In-sample edge metrics returned by `/fit` (on the realized R labels).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct InSampleMetrics {
    pub expectancy: f64,
    pub profit_factor: f64,
    pub sharpe: f64,
    pub cvar5: f64,
    pub mar: f64,
    pub n: i64,
}

/// `POST /fit` response body.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct FitResult {
    pub model_id: String,
    pub artifact_uri: String,
    pub in_sample_metrics: InSampleMetrics,
}

// --------------------------------------------------------------------------- //
// /validate
// --------------------------------------------------------------------------- //

/// CPCV fold specification. Mirrors `se_ml.contract.FoldSpec`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub struct FoldSpec {
    /// Number of contiguous time groups (>= 2).
    pub n_groups: u32,
    /// Groups held out as test per CPCV combination (>= 1).
    pub k_test_groups: u32,
    /// Bars embargoed AFTER each test block.
    pub embargo_bars: u32,
    /// Purge train labels overlapping any test span.
    pub purge: bool,
}

impl FoldSpec {
    /// Construct a [`FoldSpec`] whose `embargo_bars` and (via the horizon) purge length
    /// are derived from a [`se_core::HorizonProfile`], so labeling and cross-validation
    /// cannot desync. `purge` is always on; `n_groups`/`k_test_groups` are CV-shape choices.
    pub fn from_profile(
        profile: &se_core::HorizonProfile,
        n_groups: u32,
        k_test_groups: u32,
    ) -> Self {
        FoldSpec {
            n_groups,
            k_test_groups,
            embargo_bars: profile.embargo_bars,
            purge: true,
        }
    }
}

impl Default for FoldSpec {
    /// Matches the Python `FoldSpec()` defaults: 8 groups, 2 test, 5 embargo, purge on.
    fn default() -> Self {
        FoldSpec {
            n_groups: 8,
            k_test_groups: 2,
            embargo_bars: 5,
            purge: true,
        }
    }
}

/// `POST /validate` request body.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct MlValidateRequest {
    pub dataset_uri: String,
    pub horizon: String,
    #[serde(default)]
    pub fold_spec: FoldSpec,
    /// Number of strategy trials for DSR deflation / PBO (>= 1).
    pub n_trials: u32,
}

/// `POST /calibrate` request body.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct MlCalibrateRequest {
    pub dataset_uri: String,
    pub model_id: String,
}

/// `POST /importance` request body.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct MlImportanceRequest {
    pub dataset_uri: String,
    pub model_id: String,
}

/// Default acting threshold τ\* (0.5) for `ValidationResult::act_threshold` when a worker
/// response omits it (e.g. an older worker or a hand-built test fixture).
fn default_act_threshold() -> f64 {
    0.5
}

/// `POST /validate` response body. This is the authoritative input to the promotion gate.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ValidationResult {
    pub dsr: f64,
    pub pbo: f64,
    pub oos_expectancy_cost_aware: f64,
    pub profit_factor: f64,
    pub cvar5: f64,
    pub mar: f64,
    pub regime_contrib: BTreeMap<String, f64>,
    pub n_regimes_positive: i64,
    pub passed_gate: bool,
    /// OOS precision: fraction of OOS trades ACTED ON (prob >= τ\*) that were profitable. The
    /// north-star meta-labeling metric. Surfaced and persisted; never a ranking key here (the
    /// cost-aware OOS expectancy is already precision-conditioned upstream).
    #[serde(default)]
    pub precision_oos: f64,
    /// OOS recall at τ\*: fraction of profitable OOS trades that were acted on.
    #[serde(default)]
    pub recall_oos: f64,
    /// τ\* — the acting threshold in [0,1] the meta-label classifier acts at. Defaults to 0.5
    /// when absent so older worker responses / fixtures still deserialize.
    #[serde(default = "default_act_threshold")]
    pub act_threshold: f64,
    /// Count of OOS trades acted on at τ\* (the acted cohort size). Drives the search guardrail
    /// that refuses to promote genomes that act on too few OOS trades.
    #[serde(default)]
    pub n_acted_oos: i64,
    /// Precision on a STRICT time-ordered forward holdout: fit + threshold on the earliest 70% of
    /// rows, measure precision on the latest 30% (never shuffled). Distinguishes a durable edge
    /// from regime-fitting that shuffled CPCV folds can flatter. Reported, never a ranking key.
    #[serde(default)]
    pub precision_forward: f64,
    /// Cost-aware expectancy (R) on the same forward holdout at the forward acting threshold.
    #[serde(default)]
    pub expectancy_forward: f64,
    /// Count of acted trades in the forward holdout (small = a weak, low-confidence estimate).
    #[serde(default)]
    pub n_forward: i64,
}

// --------------------------------------------------------------------------- //
// /calibrate
// --------------------------------------------------------------------------- //

/// A single reliability-curve point.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ReliabilityPoint {
    pub predicted: f64,
    pub realized: f64,
    pub count: i64,
}

/// The fitted calibration map (e.g. isotonic).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct CalibrationMap {
    pub method: String,
    pub x: Vec<f64>,
    pub y: Vec<f64>,
}

/// `POST /calibrate` response body.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct CalibrationResult {
    pub calibration_map: CalibrationMap,
    pub reliability_points: Vec<ReliabilityPoint>,
    pub brier: f64,
}

// --------------------------------------------------------------------------- //
// /importance
// --------------------------------------------------------------------------- //

/// SHAP + permutation importance for a single feature or layer.
#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
pub struct ImportanceScore {
    pub shap: f64,
    pub permutation: f64,
}

/// `POST /importance` response body.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ImportanceResult {
    pub per_feature: BTreeMap<String, ImportanceScore>,
    pub per_layer: BTreeMap<String, ImportanceScore>,
}

File: ml-worker/src/se_ml/contract.py
"""The Rust<->Python contract: pydantic v2 job + result models.

The Rust `se-mlclient` crate mirrors these EXACT field names (snake_case). Bulk data
(features + labels) is never embedded in JSON; it is exchanged as Parquet/Arrow by
`dataset_uri` (a local path or file:// URI). Bodies carry only metadata + metrics.

JSON contract per endpoint
==========================

GET /health
    -> HealthResponse
       {"status": "ok", "version": "0.1.0"}

POST /fit
    FitRequest:
       {"dataset_uri": "/data/spy_swing.parquet",
        "horizon": "swing",
        "model_params": {"num_leaves": 31, "learning_rate": 0.05, "n_estimators": 300},
        "seed": 42}
    -> FitResult:
       {"model_id": "gbm-3f2a...",
        "artifact_uri": "/artifacts/gbm-3f2a....joblib",
        "in_sample_metrics": {"expectancy": 0.12, "profit_factor": 1.8,
                              "sharpe": 1.4, "cvar5": -0.9, "mar": 2.1, "n": 1200}}

POST /validate
    ValidateRequest:
       {"dataset_uri": "/data/spy_swing.parquet",
        "horizon": "swing",
        "fold_spec": {"n_groups": 8, "k_test_groups": 2,
                      "embargo_bars": 5, "purge": true},
        "n_trials": 50}
    -> ValidateResult:
       {"dsr": 0.61, "pbo": 0.08, "oos_expectancy_cost_aware": 0.04,
        "profit_factor": 1.35, "cvar5": -0.7, "mar": 1.2,
        "regime_contrib": {"bull": 0.06, "bear": 0.01, "chop": 0.03},
        "n_regimes_positive": 3, "passed_gate": true,
        "precision_oos": 0.58, "recall_oos": 0.41,
        "act_threshold": 0.62, "n_acted_oos": 73}

POST /calibrate
    CalibrateRequest:
       {"dataset_uri": "/data/spy_swing.parquet", "model_id": "gbm-3f2a..."}
    -> CalibrateResult:
       {"calibration_map": {"method": "isotonic",
                            "x": [...], "y": [...]},
        "reliability_points": [{"predicted": 0.1, "realized": 0.08, "count": 40}, ...],
        "brier": 0.18}

POST /importance
    ImportanceRequest:
       {"dataset_uri": "/data/spy_swing.parquet", "model_id": "gbm-3f2a..."}
    -> ImportanceResult:
       {"per_feature": {"rsi_14": {"shap": 0.21, "permutation": 0.18}, ...},
        "per_layer":   {"momentum": {"shap": 0.40, "permutation": 0.35}, ...}}
"""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from .config import DEFAULT_SEED


class _Strict(BaseModel):
    """Base model: snake_case only, forbid unexpected fields to catch contract drift."""

    model_config = ConfigDict(extra="forbid")


# --------------------------------------------------------------------------- #
# /health
# --------------------------------------------------------------------------- #
class HealthResponse(_Strict):
    status: str = "ok"
    version: str


# --------------------------------------------------------------------------- #
# /fit
# --------------------------------------------------------------------------- #
class FitRequest(_Strict):
    dataset_uri: str = Field(..., description="Parquet/Arrow path or file:// URI (features+labels)")
    horizon: str = Field(..., description="Horizon profile id, e.g. 'swing'.")
    model_params: dict[str, Any] = Field(default_factory=dict)
    seed: int = DEFAULT_SEED


class InSampleMetrics(_Strict):
    expectancy: float
    profit_factor: float
    sharpe: float
    cvar5: float
    mar: float
    n: int


class FitResult(_Strict):
    model_id: str
    artifact_uri: str
    in_sample_metrics: InSampleMetrics


# --------------------------------------------------------------------------- #
# /validate
# --------------------------------------------------------------------------- #
class FoldSpec(_Strict):
    n_groups: int = Field(8, ge=2, description="Number of contiguous time groups.")
    k_test_groups: int = Field(2, ge=1, description="Groups held out as test per CPCV combination.")
    embargo_bars: int = Field(5, ge=0, description="Bars embargoed AFTER each test block.")
    purge: bool = Field(True, description="Purge train labels overlapping any test span.")


class ValidateRequest(_Strict):
    dataset_uri: str
    horizon: str
    fold_spec: FoldSpec = FoldSpec()
    n_trials: int = Field(1, ge=1, description="Number of strategy trials (for DSR deflation).")


class ValidateResult(_Strict):
    dsr: float
    pbo: float
    oos_expectancy_cost_aware: float
    profit_factor: float
    cvar5: float
    mar: float
    regime_contrib: dict[str, float]
    n_regimes_positive: int
    passed_gate: bool
    # Precision-optimized meta-labeling acting layer (measured out-of-sample at tau*).
    precision_oos: float = Field(
        ..., description="Fraction of ACTED OOS trades that were profitable (R > 0) at tau*."
    )
    recall_oos: float = Field(
        ...,
        description="Fraction of all profitable OOS opportunities captured at tau* (coverage).",
    )
    act_threshold: float = Field(
        ..., description="tau*, the selected meta-labeling acting threshold (probability in [0,1])."
    )
    n_acted_oos: int = Field(
        ..., description="Number of OOS-reporting-half trades acted on at tau*."
    )


# --------------------------------------------------------------------------- #
# /calibrate
# --------------------------------------------------------------------------- #
class CalibrateRequest(_Strict):
    dataset_uri: str
    model_id: str


class ReliabilityPoint(_Strict):
    predicted: float
    realized: float
    count: int


class CalibrationMap(_Strict):
    method: str
    x: list[float]
    y: list[float]


class CalibrateResult(_Strict):
    calibration_map: CalibrationMap
    reliability_points: list[ReliabilityPoint]
    brier: float


# --------------------------------------------------------------------------- #
# /importance
# --------------------------------------------------------------------------- #
class ImportanceScore(_Strict):
    shap: float
    permutation: float


class ImportanceRequest(_Strict):
    dataset_uri: str
    model_id: str


class ImportanceResult(_Strict):
    per_feature: dict[str, ImportanceScore]
    per_layer: dict[str, ImportanceScore]
